src/Sprites/PaletteHeader.cpp

```diff
@@ -0,0 +1,44 @@
+#include "PaletteHeader.h"
+
+namespace PaletteTag {
+	extern const std::array<char, 4> Palette{ 'C', 'P', 'A', 'L' };
+	const std::array<char, 4> Section{ 'P', 'P', 'A', 'L' };
+	const std::array<char, 4> Header{ 'h', 'e', 'a', 'd' };
+	const std::array<char, 4> Data{ 'd', 'a', 't', 'a' };
+}
```

Comment:
Interesting how exactly one of these is declared `extern`.

I suspect this has something to do with taking the address of a constant, and the difference between declarations and definitions.

A declaration goes in the header file. It can and should define the value for constants. A definition goes in the source (.cpp) file. It does not need to define the value, which should be imported from the header file. The definition in the source file reserves actual space for the constant. As in, there are bytes written to the executable's image, and the constant has a place in the virtual address space of the process. This is what allows the address of the constant to be taken. Obviously you can't do this in the header file, or every compilation unit (.cpp) file that imports the header will allocate space for the constant, and you'd have multiple copies of it floating around in memory. That's generally a bad thing.

If you don't take the address of a constant, the compiler may be able to feed the defined constant value forward, and eliminate any references to an actual constant. Hence you can use a constant with only a declaration and no definition, and not get errors. You'd only get errors about a missing definition once you try to take the address of the constant.

Reply:
Additionally, these constants should probably be defined in the namespace of the class that uses them.

Reply:
The one defined as extern is used outside of this file. The others are only used locally. It could probably use some refactoring though.

Reply:
Used outside of this file, or defined outside of this file? I would assume that it's defined elsewhere if it's declared as `extern` here. If it's used elsewhere, that may be a problem. Either way, the asymmetry is concerning.

Using `extern` will turn a variable definition into a simple declaration. It can be used to create a forward declare to some global variable that is defined elsewhere. Generally `extern` is used in a header file. You could use `extern` from a .cpp file, as a way to import a symbol from another .cpp file without having to create a header file, and that may be suitable for a one-off import, but otherwise it would be more properly placed in a header file.

If you use `extern` in all .cpp files, you may find the variable lacks a definition, which can lead to link errors if you ever take the address of it.

----

I believe `extern` is being used incorrectly here. I've tried playing with the code a bit, but have not yet managed to demonstrate any problems.

Reply:
That variable is also defined as extern in the header file. Adding the extern in the 
.cpp file as well was a logic error on my part and I'll correct. Sorry for the confusion. We can move the variables' location out of this file when we now where we want it as well.